### PR TITLE
[JIT] Refactor SchemaCheckMode to check for mutated metadata

### PR DESCRIPTION
**Pull request by ttanpcs**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79991

- Refactored SchemaCheckMode so that arguments and outputs are only unflattened/unwrapped a single time 
- Added cloned_metadata variable inside torch_dispatch which accounts for ops that mutate stride or storage()._cdata
- Fixed aliasing check to compare with any iterable outputs (ie list of tensor output)
- Fixed aliasing check to check ops with an (a -> *) notation argument correctly
- Changed order of mutation and aliasing checks to optimize error message usefulness (generally something incorrectly aliasing implies that something is also incorrectly mutating so an aliasing error gives more information)
- Added tests for resize_, as_strided to ensure that mutated metadata checking works as intended

### CODE REVIEW

File: torch/testing/_internal/schema_check_mode.py
Comment (davidberard98) [inline torch/testing/_internal/schema_check_mode.py:95]: nit: it looks like the value will never be used if e is not a tensor... if that's true could we return `None` instead?

File: torch/testing/_internal/schema_check_mode.py
Comment (davidberard98) [inline torch/testing/_internal/schema_check_mode.py:59]: cc @eellison just to make sure this is the right logic for wildcards

**Review by davidberard98 — APPROVED** [review]
looks good! just wanted to ask @eellison to take a look at the wildcard logic so please wait for his review first

(also thanks for adding the wildcard stuff, I had almost forgotten about it)

File: torch/testing/_internal/schema_check_mode.py
Comment (Gamrix) [inline torch/testing/_internal/schema_check_mode.py:159]: Separate this if statement out into multiple lines, and give 
`bool(len(out_alias_pairs_map[i] & out_alias_pairs_map[j]))` a variable name

File: torch/testing/_internal/schema_check_mode.py
Comment (Gamrix) [reply torch/testing/_internal/schema_check_mode.py:59]: You still need a check that the types are the same. Wildcard aliases do not alias objects of different type classes. See `alias_analysis.h:40`.

File: torch/testing/_internal/schema_check_mode.py
Comment (Gamrix) [inline torch/testing/_internal/schema_check_mode.py:48]: Why did you change this from an `isinstance()` check to instead check type equality? 

Also, this line is very long and should just be a normal if statement.

Comment (goldenxuett) [reply torch/testing/_internal/schema_check_mode.py:48]: This line was changed from isinstance() to check type equality because it is possible for a subclass of torch.Tensor to pass through the unwrapping due to not having an "elem" instance variable which will then error out on the subsequent mutation check / aliasing check. Changing the checks to type equality ensures this doesn't happen.

File: torch/testing/_internal/schema_check_mode.py
Comment (goldenxuett) [reply torch/testing/_internal/schema_check_mode.py:95]: Yup, I had this from an old version where I was running into issues with having None instead, but it seems to work fine now

File: torch/testing/_internal/schema_check_mode.py
Comment (Gamrix) [reply torch/testing/_internal/schema_check_mode.py]: Do you have some example code handy for where this would happen? I assume that we want to still check for aliasing information for tensor subclasses. 

(Also, this is good stuff to add as a comment to this function).

Comment (goldenxuett) [reply torch/testing/_internal/schema_check_mode.py]: It seems to happen with a "Parameter" object which has no attribute 'elem' when calling the batch norm op. For all other tensor subclasses which have a tensor within them, aliasing information is checked as expected.

File: torch/testing/_internal/schema_check_mode.py
Comment (bdhirsh) [inline torch/testing/_internal/schema_check_mode.py:90]: Is the idea here that you want to assert that both the tensor data and metadata are unchanged after performing an op?

Are you able to use the `assertEqual()` method from the testing framework ([here](https://github.com/pytorch/pytorch/blob/09f79e94acce30e2002063322c9fe50d116361f1/torch/testing/_internal/common_utils.py#L2210)) instead? It should already handle checking metadata for you.

Another reason to use the one from the framework is that keeping track of the full set of metadata to check is tough, and can change over time. It looks like today, that method also checks `layout` and `device`: https://github.com/pytorch/pytorch/blob/09f79e94acce30e2002063322c9fe50d116361f1/torch/testing/_comparison.py#L635.

Comment (goldenxuett) [reply torch/testing/_internal/schema_check_mode.py:90]: From what I understand, the issue with assertEqual() is that torch.clone() doesn't copy the strides and storage cdata, so that assertEqual might return true when the strides have in fact changed or vice versa. Thus the parse_metadata function is there specifically to record data that isn't copied correctly in torch.clone().

Comment (davidberard98) [reply torch/testing/_internal/schema_check_mode.py:90]: I agree with @goldenxuett , I think we need a separate check for these, because `clone()` doesn't copy strides or layout, and so we need to check those properties separately from the `assertEquals`

File: torch/testing/_internal/schema_check_mode.py
Comment (eellison) [reply torch/testing/_internal/schema_check_mode.py:61]: The semantics of `chunk(Tensor(a -> *) self, int chunks, int dim=0) -> Tensor(a)[]` is that the input tensor storage will be contained in the output list

**Review by davidberard98 — APPROVED** [review]
LGTM! You might need to rebase before it lets you merge it.

**Pull request closed without merging**
